enhancements/run-bundle.md

```diff
@@ -0,0 +1,263 @@
+---
+title: run-bundle
+authors:
+  - "@jmrodri"
+reviewers:
+  - "@joelanford"
+  - "@estroz"
+  - "@kevinrizza"
+  - "@ecordell"
+  - "@shawn-hurley"
+approvers:
+  - "@joelanford"
+  - "@estroz"
+  - "@ecordell"
+  - "@shawn-hurley"
+creation-date: 2020-05-07
+last-updated: 2020-05-07
+status: implementable
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Run Bundle
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [X] Test plan is defined
+- [X] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions
+
+1. We currently check OLM installation status with `operator-sdk olm status`.
+   Can we change that to using the discovery API?
+   * Based on discussions, using the discovery API is the way to go.
+
+2. How do we handle --mode= flag on registry add
+   * should we ALWAYS default to --mode=semver?
+     * this has less testing and isnot yet the default. We should stick with
+        the default.
+   * should we have a flag that allows our default mode handling to be overridden?
+     * no
```

Comment:
What is the reasoning is here? I imagine either:

- you would want to replicate the mode you intend to publish with, so that you can see upgrade behavior
- you don't care about mode at all because `run bundle` always runs without the context of a full index (i.e. always builds an index of one operator). if that's the case we can short circuit the mode discussion because it doesn't make a difference.

Reply:
The index is potentially not full, since a user can pass an arbitrary index image (or none) to the command.

Reply:
Making sure we're on the same page - this command:

- takes an index
- adds a bundle to it
- runs the bundle that it added to the index

If that's the case, then I think users may care how their bundle is added to the index, right? `opm` doesn't help in this case, because it's not in the `add` loop here.

Reply:
That is a correct assessment of what the command does.

>I think users may care how their bundle is added to the index

Is this variance a blocker for an MVP?

>opm doesn't help in this case, because it's not in the add loop here

Are `opm registry add` (what is happening internally in `run bundle`) and `opm index add` (what a user can do externally) different enough that `opm` does not help, or are you talking about the internal `opm registry add` invocation not being configurable?

Reply:
> Is this variance a blocker for an MVP?

I don't know that it's a blocker, but it doesn't seem like a huge headache to plumb the configuration through. It's just easy to predict the bug report "bundle run produces different behavior than when my operator is published".

> are you talking about the internal opm registry add invocation not being configurable?

Yes, that's all - this thread is on whether the mode should be configurable. I'm pointing out that `opm` is not an escape hatch for this case, because `opm` won't do the "run bundle" part (i.e. you get to choose between good testing ux provided by `run bundle` or accurate testing via more painful `opm` steps).

Reply:
Based on discussions at Group F meeting, this is what we ended up with:

2. How do we handle --mode= flag on registry add?
   * When adding a bundle to an empty index, default to **semver**
   * When adding a bundle to a provided index, default to **replaces**
   * Should the add mode be configurable on the CLI? **optional flag**;
     not in an example right now